### PR TITLE
Add events

### PR DESCRIPTION
**Pull request by mbraak**
Make it possible to receive events. Preliminary api:

```js
import EventEmitter from 'eventemitter3';

const eventEmitter = new EventEmitter();

eventEmitter.on('addUpload', ({ element, fieldName, upload }) => {
  console.log('addUpload', element, fieldName, upload);
});

eventEmitter.on('removeUpload', ({ element, fieldName, upload }) => {
  console.log('removeUpload', element, fieldName, upload);
});


initUploadFields(
  document.getElementById("example-form"),
  {
    eventEmitter,
  }
);
```

### CODE REVIEW

**Comment by BoPeng** [comment]
This is great. I will add the Python part (`self.meta`) to this PR and see if I can implement the checkbox this way.

**Comment by BoPeng** [comment]
This is now #343

**Comment by mbraak** [comment]
Pr #343 is merged into this one. This pr will be merged shortly.

File: frontend/src/upload_file.ts [hunk 1]
```diff
@@ -240,9 +262,18 @@ class UploadFile {
...
     upload.start();
-    renderer.addNewUpload(filename, uploadIndex);
+
+    const element = renderer.addNewUpload(filename, uploadIndex);
     this.uploads.push(upload);
     this.uploadStatuses.push("uploading");
+
+    if (this.eventEmitter) {
+      this.eventEmitter.emit("addUpload", {
+        element,
```
Comment (BoPeng) [inline frontend/src/upload_file.ts:272]: Did you forgot to pass `metaDataField` here?

File: frontend/src/upload_file.ts [hunk 2]
```diff
@@ -276,6 +307,14 @@ class UploadFile {
       return;
     }
 
+    if (this.eventEmitter) {
+      this.eventEmitter.emit("removeUpload", {
+        element: this.renderer.findFileDiv(uploadIndex),
+        fieldName: this.fieldName,
```
Comment (BoPeng) [inline frontend/src/upload_file.ts:313]: Also missing `metaDataField`

File: frontend/src/upload_file.ts (hunk 1 above)
Comment (mbraak) [reply frontend/src/upload_file.ts:268]: fixed

File: frontend/src/upload_file.ts (hunk 2 above)
Comment (mbraak) [reply frontend/src/upload_file.ts:310]: fixed

**Comment by BoPeng** [comment]
I can see that `addUpload` is emitted when

1. form is initialized, when `upload` is `UploadedFile`
2. when a file is uploaded when `upload` is `Upload` or `S3Uploader`.

A side effect of this is that the passed `upload` does not have a uniform interface. For example, `upload.name` does not always exist (e.g. for `Upload`, which should be `upload.file.name`) so the user code will need to check the type of `upload` or existence of `upload.name` , or even probe into `element` to get the filename.

Since filename is important for the use of `metaDataField`, can we do something here?

**Comment by mbraak** [comment]
Good point. I can add a `fileName` parameter.

The best solution would be to give the different uploads types the same interface, but that is something for later.

**Comment by BoPeng** [comment]
Thanks for the addition of `fileName`. I can confirm that this PR works well and I can use it to add two widgets to the frontend. The only catch was that 

```
element.appendChild(descElem)
```
adds the input box after the filename during upload (when `filesize` etc are unabailable), and after `filesize` and delete button when the files are loaded as placeholder. It can be slightly better to change this line to append after the first child (filename), but really should not matter for an example.

**Comment by mbraak** [comment]
Changed in the example: the input is now added after the filename

**Comment by BoPeng** [comment]
Cool, the last missing piece would be the description of API in `README.md`....

**Comment by mbraak** [comment]
I added documentation. I added a note that the events are an experimental feature. Meaning that the api can change.

**Pull request merged**